demos/common/processing/async_pipeline.cpp

```diff
@@ -0,0 +1,217 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+*/
+
+#include "async_pipeline.h"
+#include <samples/args_helper.hpp>
+#include <cldnn/cldnn_config.hpp>
+#include <samples/slog.hpp>
+
+using namespace InferenceEngine;
+
+PipelineBase::PipelineBase(ModelBase* modelInstance, const CnnConfig& cnnConfig, InferenceEngine::Core* engine){
+    model.reset(modelInstance);
+
+    auto ie = engine ?
+        std::unique_ptr<InferenceEngine::Core, void(*)(InferenceEngine::Core*)>(engine, [](InferenceEngine::Core*) {}) :
+        std::unique_ptr<InferenceEngine::Core, void(*)(InferenceEngine::Core*)>(new InferenceEngine::Core, [](InferenceEngine::Core* ptr) {delete ptr; });
+
+    // --------------------------- 1. Load inference engine ------------------------------------------------
+    slog::info << "Loading Inference Engine" << slog::endl;
+
+    slog::info << "Device info: " << slog::endl;
+    std::cout << ie->GetVersions(cnnConfig.devices);
+
+    /** Load extensions for the plugin **/
+    if (!cnnConfig.cpuExtensionsPath.empty()) {
+        // CPU(MKLDNN) extensions are loaded as a shared library and passed as a pointer to base extension
+        IExtensionPtr extension_ptr = make_so_pointer<IExtension>(cnnConfig.cpuExtensionsPath.c_str());
+        ie->AddExtension(extension_ptr, "CPU");
+    }
+    if (!cnnConfig.clKernelsConfigPath.empty()) {
+        // clDNN Extensions are loaded from an .xml description and OpenCL kernel files
+        ie->SetConfig({ {PluginConfigParams::KEY_CONFIG_FILE, cnnConfig.clKernelsConfigPath} }, "GPU");
+    }
+
+    // --------------------------- 2. Read IR Generated by ModelOptimizer (.xml and .bin files) ------------
+    slog::info << "Loading network files" << slog::endl;
+    /** Read network model **/
+    InferenceEngine::CNNNetwork cnnNetwork = ie->ReadNetwork(model->getModelFileName());
+    /** Set batch size to 1 **/
+    slog::info << "Batch size is forced to 1." << slog::endl;
+    cnnNetwork.setBatchSize(1);
+
+    // -------------------------- Reading all outputs names and customizing I/O blobs (in inherited classes)
+    model->prepareInputsOutputs(cnnNetwork);
+
+    // --------------------------- 4. Loading model to the device ------------------------------------------
+    slog::info << "Loading model to the device" << slog::endl;
+    execNetwork = ie->LoadNetwork(cnnNetwork, cnnConfig.devices, cnnConfig.execNetworkConfig);
+
+    // --------------------------- 5. Create infer requests ------------------------------------------------
+    requestsPool.reset(new RequestsPool(execNetwork, cnnConfig.maxAsyncRequests));
+
+    // --------------------------- 6. Call onLoadCompleted to complete initialization of model -------------
+    model->onLoadCompleted(&execNetwork, requestsPool.get());
```

Comment:
This should be done in model constructor. The only thing that could depend on the pipeline is batch size. But it is advised to use throughput mode instead of batching.

Reply:
That's questionable. Id' keep it here. If we move to model, we'll have to pass core pointer there, etc. It's better to keep everything in one place (in pipeline) and move only those parts of code that are different from each other to the model classes.

Reply:
Ok, I can't come up better solution.